trustmanager/keyfilestore.go

```diff
@@ -106,64 +110,122 @@ func (s *KeyMemoryStore) ListKeys() []string {
 	pemPrivKey, err := KeyToPEM(privKey)
 	if err != nil {
 		return err
 	}
 
-	return s.Add(name, pemPrivKey)
-}
-
-func getKey(s LimitedFileStore, name string) (data.PrivateKey, error) {
-	keyBytes, err := s.Get(name)
-	if err != nil {
-		return nil, err
+	attempts := 0
+	passphrase := ""
+	giveup := false
+	for {
+		passphrase, giveup, err = passphraseRetriever(name, alias, true, attempts)
+		if err != nil {
+			attempts++
+			continue
+		}
+		if giveup {
+			return errors.New("obtaining passphrase failed")
+		}
+		if attempts > 10 {
+			return errors.New("sanity check on number of passphrase attempts exceeded")
+		}
+		break
 	}
 
-	// Convert PEM encoded bytes back to a PrivateKey
-	privKey, err := ParsePEMPrivateKey(keyBytes, "")
-	if err != nil {
-		return nil, err
+	if passphrase != "" {
+		pemPrivKey, err = EncryptPrivateKey(privKey, passphrase)
+		if err != nil {
+			return err
+		}
 	}
 
-	return privKey, nil
+	return s.Add(name+"_"+alias, pemPrivKey)
 }
 
-func addEncryptedKey(s LimitedFileStore, name string, privKey data.PrivateKey, passphrase string) error {
-	encryptedPrivKey, err := EncryptPrivateKey(privKey, passphrase)
-	if err != nil {
-		return err
+func getKeyAlias(s LimitedFileStore, keyID string) (string, error) {
+	files := s.ListFiles(true)
+	name := strings.TrimSpace(strings.TrimSuffix(filepath.Base(keyID), filepath.Ext(keyID)))
+
+	for _, file := range files {
+		lastPathSeparator := strings.LastIndexAny(file, "/\\")
+		filename := file[lastPathSeparator+1:]
+
+		if strings.HasPrefix(filename, name) {
+			aliasPlusDotKey := strings.TrimPrefix(filename, name+"_")
+			retVal := strings.TrimSuffix(aliasPlusDotKey, "."+keyExtension)
+			return retVal, nil
+		}
 	}
 
-	return s.Add(name, encryptedPrivKey)
+	return "", fmt.Errorf("keyId %s has no alias", name)
 }
 
-func getDecryptedKey(s LimitedFileStore, name string, passphrase string) (data.PrivateKey, error) {
-	keyBytes, err := s.Get(name)
+// GetKey returns the PrivateKey given a KeyID
+func getKey(s LimitedFileStore, passphraseRetriever PassphraseRetriever, name string) (data.PrivateKey, error) {
+	keyAlias, err := getKeyAlias(s, name)
 	if err != nil {
 		return nil, err
 	}
 
-	// Gets an unencrypted PrivateKey.
-	privKey, err := ParsePEMPrivateKey(keyBytes, passphrase)
+	keyBytes, err := s.Get(name + "_" + keyAlias)
 	if err != nil {
 		return nil, err
 	}
 
+	// See if the key is encrypted. If its encrypted we'll fail to parse the private key
+	privKey, err := ParsePEMPrivateKey(keyBytes, "")
+	if err != nil {
+		// We need to decrypt the key, lets get a passphrase
+		attempts := 0
+		for {
+			passphrase, giveup, err := passphraseRetriever(name, string(keyAlias), false, attempts)
+			// Check if the passphrase retriever got an error or if it is telling us to give up
+			if giveup || err != nil {
+				return nil, errors.New("obtaining passphrase failed")
+			}
+			if attempts > 10 {
```

Comment:
Same comment as above, if user might get asked for password on every call, 10 attempts seems abusive.